src/tests/Microsoft.PowerFx.Interpreter.Tests/LanguageServiceProtocol/LanguageServerTests.cs

```diff
@@ -647,6 +649,99 @@ public void TestPublishTokens()
+        [InlineData("{}", "\"hi\"", typeof(StringType))]
+        [InlineData("{}", "", typeof(BlankType))]
+        [InlineData("{}", "{ A: 1 }", typeof(RecordType))]
+        [InlineData("{}", "[1, 2, 3]", typeof(TableType))]
+        [InlineData("{}", "true", typeof(BooleanType))]
+        public void TestPublishExpressionType(string context, string expression, System.Type expectedType)
+        {
+            var documentUri = $"powerfx://app?context={context}&getExpressionType=true";
+            _testServer.OnDataReceived(JsonSerializer.Serialize(new
+            {
+                jsonrpc = "2.0",
+                method = "textDocument/didOpen",
+                @params = new DidOpenTextDocumentParams()
+                {
+                    TextDocument = new TextDocumentItem()
+                    {
+                        Uri = documentUri,
+                        LanguageId = "powerfx",
+                        Version = 1,
+                        Text = expression
+                    }
+                }
+            }));
+
+            Assert.Equal(2, _sendToClientData.Count);
+            var response = JsonSerializer.Deserialize<JsonRpcPublishExpressionTypeNotification>(_sendToClientData[1], _jsonSerializerOptions);
+            Assert.Equal("$/publishExpressionType", response.Method);
+            Assert.Equal(documentUri, response.Params.Uri);
+            Assert.IsType(expectedType, response.Params.Type);
+        }
+
+        [Theory]
+        [InlineData("{\"A\": 1 }", "invalid+A")]
+        [InlineData("{}", "B")]
+        [InlineData("{}", "+")]
+        public void TestPublishExpressionType_Null(string context, string expression)
+        {
+            var documentUri = $"powerfx://app?context={context}&getExpressionType=true";
+            _testServer.OnDataReceived(JsonSerializer.Serialize(new
+            {
+                jsonrpc = "2.0",
+                method = "textDocument/didOpen",
+                @params = new DidOpenTextDocumentParams()
+                {
+                    TextDocument = new TextDocumentItem()
+                    {
+                        Uri = documentUri,
+                        LanguageId = "powerfx",
+                        Version = 1,
+                        Text = expression
+                    }
+                }
+            }));
+
+            Assert.Equal(2, _sendToClientData.Count);
+            var response = JsonSerializer.Deserialize<JsonRpcPublishExpressionTypeNotification>(_sendToClientData[1], _jsonSerializerOptions);
+            Assert.Equal("$/publishExpressionType", response.Method);
+            Assert.Equal(documentUri, response.Params.Uri);
+            Assert.Null(response.Params.Type);
+        }
+
+        [Theory]
+        [InlineData("{}", "{ A: 1 }", "{\"$type\":\"Record\",\"A\":{\"$type\":\"Number\"}}")]
+        [InlineData("{}", "[1, 2]", "{\"$type\":\"Table\",\"Value\":{\"$type\":\"Number\"}}")]
```

Comment:
What about `{'$type': 123}`?

Reply:
also, nested records, `{A: 1, B: { C: { D: "Qwerty" }, E: true } } `

Reply:
> also, nested records, {A: 1, B: { C: { D: "Qwerty" }, E: true } }

Nested records serialize the whole shape. Your example would be
```json
"{\"$type\":\"Record\",\"A\":{\"$type\":\"Number\"},\"B\":{\"$type\":\"Record\",\"C\":{\"$type\":\"Record\",\"D\":{\"$type\":\"String\"}},\"E\":{\"$type\":\"Boolean\"}}}"
```
I can add another test for nested records.

Reply:
> What about {'$type': 123}?

This is currently undefined and produces duplicate `$type` keys in serialization, and that string then fails to deserialize. I considered adding another layer of nesting to avoid these conflicts, but I prefer the flattened interface.

I can update this to exclude `$type` from being serialized from the expression result, if you agree. Otherwise I can nest the type shape as well, something like

```json
{
  "type": "Record",
  "names": { "type": { /* Type from expression's `type` }}
}
```

Reply:
I'd lean towards nested, since `{'$type': 123}` is a valid expression. Especially if we ever want to add more keys to this, it'll be easier if we don't have to worry about name conflicts.